collections/nemo_nlp/nemo_nlp/data/datasets/utils.py

```diff
@@ -312,6 +317,106 @@ def process_atis(infold, uncased, modes=['train', 'test'], dev_split=0):
+        if start_i == tag_start:
+            return f'B-{tag_str}'
+        elif start_i > tag_start and end_i <= tag_end:
+            return f'I-{tag_str}'
+    return "O"
+
+
+def process_triple_datasets(infold, uncased, dataset_name,
+                            modes=['train', 'test', 'eval'], dev_split=0):
+    """ process and convert triple datasets into NeMo's BIO format
+    """
+
+    from nltk.tokenize import WhitespaceTokenizer
+
+    outfold = f'{infold}/{dataset_name}-nemo-processed'
+    infold = f'{infold}/'
+
+    if uncased:
+        outfold = f'{outfold}-uncased'
+
+    if if_exist(outfold, [f'{mode}.tsv' for mode in modes]):
+        logger.info(LOGGING_TMP.format(dataset_name, outfold))
+        return outfold
+    logger.info(f'Processing {dataset_name} dataset and store at {outfold}')
+
+    os.makedirs(outfold, exist_ok=True)
+
+    outfiles = {}
+    intents_list = {}
+    slots_list = {}
+    slots_list_all = {}
+
+    outfiles['dict_intents'] = open(f'{outfold}/dict.intents.csv', 'w')
+    outfiles['dict_slots'] = open(f'{outfold}/dict.slots.csv', 'w')
+
+    outfiles['dict_slots'].write('O\n')
+    slots_list["O"] = 0
+    slots_list_all["O"] = 0
+
+    for mode in modes:
+        outfiles[mode] = open(os.path.join(outfold, mode + '.tsv'), 'w')
+        outfiles[mode].write('sentence\tlabel\n')
+        outfiles[mode + '_slots'] = open(f'{outfold}/{mode}_slots.tsv', 'w')
+
+        queries = open(f'{infold}/{mode}.tsv', 'r').readlines()
+
+        for i, query in enumerate(queries):
+            [intent_str, slot_tags_str, sentence] = query.strip().split("\t")
+
+            if intent_str not in intents_list:
+                intents_list[intent_str] = len(intents_list)
+                outfiles['dict_intents'].write(f'{intent_str}\n')
+
+            outfiles[mode].write(f'{sentence[4:-4]}\t'
+                                 f'{str(intents_list[intent_str])}\n')
+
+            if str.count(slot_tags_str, ":") > 0:
+                slot_tags = slot_tags_str.strip().split(",")
+                slot_tags_list = []
+                for st in slot_tags:
+                    if len(st.strip()) == 0:
+                        continue
+                    [start_i, end_i, slot_name] = st.strip().split(":")
+                    slot_tags_list.append([int(start_i),
+                                           int(end_i),
+                                           slot_name])
+                    if slot_name not in slots_list:
+                        slots_list[slot_name] = len(slots_list)
+                        slots_list_all[f'B-{slot_name}'] = len(slots_list_all)
+                        slots_list_all[f'I-{slot_name}'] = len(slots_list_all)
+                        outfiles['dict_slots'].write(f'B-{slot_name}\n')
+                        outfiles['dict_slots'].write(f'I-{slot_name}\n')
+            else:
+                slot_tags_list = []
+
+            span_generator = WhitespaceTokenizer().span_tokenize(sentence)
```

Comment:
I think this entire function can be rewritten to be a lot more efficient and easier to understand. So the slots are indexed like this: 
start0:end0:label0, start1:end1:label1, ...

So for each sentence, we can do this:
- start at index 0. all words from index 0 to index start0 is tagged with 'O'
- for each chunk starti:endi: the first word is add B-tag, the rest is added I-tag
- for each chunk between endi:starti+1, each word is labeled with 'O'

Reply:
I can also merge your PR as is and rewrite the function

Reply:
Seems a good idea. I can rewrite it later more efficiently without WhitespaceTokenizer.